File: model/src/main/java/net/qutester/model/step/BasicStepDef.kt
package net.qutester.model.step

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import net.qutester.model.enums.StepPhaseEnum
import net.qutester.model.infrastructure.path.Path
import net.qutester.model.text.StepPattern
import net.qutester.model.text.parts.ParamStepPatternPart
import net.qutester.model.warning.Warning

data class BasicStepDef @JsonCreator constructor(@JsonProperty("phase") override val phase: StepPhaseEnum,
                                                 @JsonProperty("stepPattern") override val stepPattern: StepPattern,
                                                 @JsonProperty("description") override val description: String? = null,
                                                 @JsonProperty("tags") override val tags: List<String> = emptyList(),
                                                 @JsonProperty("className") val className: String,
                                                 @JsonProperty("methodName") val methodName: String,
                                                 @JsonProperty("warnings") override val warnings: List<Warning> = emptyList()): StepDef {

    private val _path = run {
        val paramTypes = stepPattern.patternParts
                .filter { it is ParamStepPatternPart }
                .map { (it as ParamStepPatternPart).type }

        val packagesWithClass: List<String> = className.split('.')
        val packages: List<String> = packagesWithClass.dropLast(1)
        val classNameWithoutPackages = packagesWithClass.last();

        return@run Path(packages, classNameWithoutPackages, methodName +"(" + paramTypes.joinToString(",")+")")
    }

    override val path: Path
        get() = _path

    private val _id: String = _path.toString()

    override val id: String
        get() = _id

    @get:JsonProperty("descendantsHaveWarnings")
    override val descendantsHaveWarnings: Boolean
        get() = false // basic steps don't have descendants

    override fun toString() = buildString {
        append(phase)
        append(" ")
        append(stepPattern.toDebugString(varPrefix = "<<", varSuffix = ">>"))
    }

}

File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/testerum-model/model/step/def/basic-step-def.ts
import {StepPhaseEnum} from "./step-phase-enum";
import {StepPattern} from "./pattern/step-pattern";
import {MarshallingUtils} from "../../../json-marshalling/marshalling-utils";
import {StepDef} from "./step-def";
import {Path} from "../../path";

export class BasicStepDef implements StepDef {

    constructor(public readonly id: string,
                public readonly path: Path,
                public readonly phase: StepPhaseEnum,
                public readonly stepPattern: StepPattern,
                public readonly description: string | null,
                public readonly tags: Array<string>,
                public readonly className: string,
                public readonly methodName: string) { }

    static parse(input: Object): BasicStepDef {
        const id = input["id"];
        const path = Path.parse(input["path"]);
        const phase = MarshallingUtils.parseEnum(input["phase"], StepPhaseEnum);
        const stepPattern = StepPattern.parse(input["stepPattern"]);
        const description = input["description"];
        const tags = MarshallingUtils.parseListOfStrings(input["tags"]);
        const className = input["className"];
        const methodName = input["methodName"];

        return new BasicStepDef(id, path, phase, stepPattern, description, tags, className, methodName);
    }

}
